### PR TITLE
fromTensor: full proof works, two believed sorrys are to be filled in

### PR DESCRIPTION
**Pull request by bollu**
Goal: clear the `sorry`s for the definition of fromTensor. Pairing with @goens .

### CODE REVIEW

File: SSA/Projects/FullyHomomorphicEncryption/Basic.lean [hunk 1]
```diff
@@ -485,11 +495,57 @@ theorem R.fromTensor_eq_fromTensor'_fromPoly {q n} : R.fromTensor (q := q) (n :=
...
   support, coefficients, etc. -/
 noncomputable def R.fromTensorFinsupp (coeffs : List ℤ) : (ZMod q)[X] :=
   Polynomial.ofFinsupp (List.toFinsupp (coeffs.map Int.cast))
 
+theorem Polynomial.degree_toFinsupp [Semiring M] [DecidableEq M]
+  (xs : List M) :
+  degree { toFinsupp := List.toFinsupp (l := xs) } ≤ List.length xs := by
+    cases xs
+    case nil =>
+      simp[degree]
```
Comment (goens) [inline SSA/Projects/FullyHomomorphicEncryption/Basic.lean:509]: nit: whitespace (and all the others below)

File: SSA/Projects/FullyHomomorphicEncryption/Basic.lean [hunk 2]
```diff
@@ -534,44 +590,37 @@ theorem R.coeff_fromPoly {q n : ℕ} [Fact (q > 1)] (p : (ZMod q)[X]) : R.coeff
   have H := R.representative_fromPoly_toFun (a := p) (n := n)
   norm_cast at H ⊢
 
-
-/-- since `0` is the additive identity, getting from `(x :: xs)` at index `i` with -/
-theorem List.getD_snoc_zero (x : ℤ) (xs : List ℤ) :
-  (xs ++ [x]).getD i 0 = (xs.getD i 0) + (if i == xs.length then x else 0) := by sorry
-
+/- 1. given a list of coefficients, building the ring element and then picking the repr from
```
Comment (goens) [inline SSA/Projects/FullyHomomorphicEncryption/Basic.lean:593]: this was our map of the proof, I guess we don't need it anymore? (seems a bit confusing here with context)

File: SSA/Projects/FullyHomomorphicEncryption/Basic.lean [hunk 3]
```diff
@@ -534,44 +590,37 @@ theorem R.coeff_fromPoly {q n : ℕ} [Fact (q > 1)] (p : (ZMod q)[X]) : R.coeff
...
+  rw[coeff]
+  rw[representative_fromPoly_eq]
+  apply fromTensorFinsupp_coeffs
+  case DEGREE =>
+    generalize htensor_degree : degree (fromTensorFinsupp q tensor) = tensor_degree
+    rw[f_deg_eq]
+    cases tensor_degree
+    case none => norm_cast; apply WithBot.none_lt_some
+    case some tensor_degree =>
+      /- I hate this coercion stuff -/
```
Comment (goens) [inline SSA/Projects/FullyHomomorphicEncryption/Basic.lean:614]: I feel ya!

**Comment by goens** [comment]
LGTM

**Review by goens — APPROVED** [review]

File: SSA/Projects/FullyHomomorphicEncryption/Basic.lean (hunk 1 above)
Comment (bollu) [reply SSA/Projects/FullyHomomorphicEncryption/Basic.lean:509]: @goens what exactly is the nit?

Comment (bollu) [reply SSA/Projects/FullyHomomorphicEncryption/Basic.lean:509]: oh I see between the parens, OK

**Pull request merged**